File: minesweeper.py
import random
from itertools import product # 多重ループを減らす
import numpy as np

MS_SIZE = 8          # ゲームボードのサイズ
CLOSE, OPEN, FLAG = 0, 1, 2

class Game:

    def __init__(self, number_of_mines = 10):
        """ ゲームボードの初期化
        
        Arguments:
        number_of_mines -- 地雷の数のデフォルト値は10

        Side effects:
        mine_map[][] -- 地雷マップ(-1: 地雷，>=0 8近傍の地雷数)
        game_board[][] -- 盤面 (0: CLOSE(初期状態), 1: 開いた状態, 2: フラグ)

        """

        self.init_game_board()
        self.init_mine_map(number_of_mines)
        self.count_mines()

    def init_game_board(self):
        """ ゲーム盤を初期化 """

        self.game_board = np.array([[CLOSE for i in range(MS_SIZE)] for j in range(MS_SIZE)])              

    def init_mine_map(self, number_of_mines):
        """ 地雷マップ(self->mine_map)の初期化
        Arguments:
        number_of_mines -- 地雷の数
        
        地雷セルに-1を設定する．      
        """

        self.mine_map = np.array([[CLOSE for i in range(MS_SIZE)] for j in range(MS_SIZE)])

        # 例外処理
        if number_of_mines < 0:
        	number_of_mines = 0
        elif number_of_mines > 64:
        	number_of_mines = 64


        fy = np.array([i for i in range(MS_SIZE**2)]) # 0~63の行列の用意する
        end = MS_SIZE ** 2 # 最後位

        for i in range(number_of_mines):
        	m = random.choice(range(0,end)) # 0~end-1までランダムに数値を取り出す
        	y = int(fy[m] / MS_SIZE) # それに対応する2次元配列の座標に変換する
        	x = fy[m] % MS_SIZE
        	self.mine_map[y][x] = -1 # 地雷として設定する
        	fy[m] = fy[end - 1] # 一番後ろの数値ともうすでに取り出した数値と書き換える
        	end -= 1 # 最後位を更新する

    def count_mines(self):
        """ 8近傍の地雷数をカウントしmine_mapに格納 
        地雷数をmine_map[][]に設定する．
        """
		
        for y in range(MS_SIZE):
        	for x in range(MS_SIZE):
        		if self.mine_map[y][x] == -1:
        			pass
        		else:
        			# 例外処理
        			y_lower = 0 if y-1<0 else y-1
        			y_upper = MS_SIZE if y+2>MS_SIZE else y+2
        			x_lower = 0 if x-1<0 else x-1
        			x_upper = MS_SIZE if x+2>MS_SIZE else x+2

        			self.mine_map[y][x] = np.sum(self.mine_map[y_lower:y_upper, x_lower:x_upper] == -1) # 地雷数をカウントする

    def open_cell(self, x, y):
        """ セル(x, y)を開ける
        Arguments:
        x, y -- セルの位置
        
        Returns:
          True  -- 8近傍セルをOPENに設定．
                   ただし，既に開いているセルの近傍セルは開けない．
                   地雷セル，FLAGが設定されたセルは開けない．
          False -- 地雷があるセルを開けてしまった場合（ゲームオーバ）
        """

        row = [-1, 0, 1]
  
        if self.mine_map[y][x] == -1: # 地雷を選んだ場合ゲーム終了
       		return False
       	elif self.game_board[y][x] == OPEN: # もうすでに開けてるセルを選んだ場合何も処理しない
       		pass
       		return True
       	elif self.game_board[y][x] == FLAG: # flagのところを選んだらセルを開ける状態に切り替える
       		self.game_board[y][x] = OPEN
       	elif self.game_board[y][x] == CLOSE: # closeのセルを開ける
       		self.game_board[y][x] = OPEN
       	
       	for j, i in product(row, row):
       		if y+j<0 or x+i<0 or y+j>7 or x+i>7: # 例外処理
       			pass
       		elif i == 0 and j == 0:
       			pass
       		else:
        		if self.game_board[y + j][x + i] == OPEN: # 既に開いているセルの近傍セルは開けない
        			pass
        		elif self.game_board[y + j][x + i] == FLAG:	# FLAGが設定されたセルは開けない
       				pass
       			elif self.game_board[y + j][x + i] == CLOSE:
       				if self.mine_map[y + j][x + i] == -1: # 地雷セルは開けない
       					pass
       				elif self.mine_map[y + j][x + i] >= 0:
       					self.game_board[y + j][x + i] = OPEN	

        return True
    
    def flag_cell(self, x, y):
        """
        セル(x, y)にフラグを設定する，既に設定されている場合はCLOSE状態にする
        """

        if self.game_board[y][x] == FLAG:
        	self.game_board[y][x] = CLOSE
        elif self.game_board[y][x] == CLOSE:
        	self.game_board[y][x] = FLAG
        else:
        	self.game_board[y][x] = OPEN
            
    def is_finished(self):
        """ 地雷セル以外のすべてのセルが開かれたかチェック """
		
        for y, x in product(range(MS_SIZE), range(MS_SIZE)):
        	if not((self.game_board[y][x] != OPEN and self.mine_map[y][x] == -1) or (self.game_board[y][x] == OPEN and self.mine_map[y][x] >= 0)):
        		return False

        return True
       
    def print_header(self):
        print("=====================================")
        print("===  Mine Sweeper Python Ver. 1  ====")
        print("=====================================")

    def print_footer(self):
        print("   ", end="")
        for x in range(MS_SIZE):
            print("---", end="")
        print("[x]\n   ", end="")
        for x in range(MS_SIZE):
            print("%3d"%x, end="")
        print("")
        
    def print_mine_map(self):
        print(" [y]")
        for y in range(MS_SIZE):
            print("%2d|"%y, end="")
            for x in range(MS_SIZE):
                print("%2d"%self.mine_map[y][x], end="")
            print("")
        
    def print_game_board(self):
        marks = ['x', ' ', 'P']
        self.print_header()
        print("[y]")
        for y in range(MS_SIZE):
            print("%2d|"%y, end="")
            for x in range(MS_SIZE):
                if self.game_board[y][x] == OPEN and self.mine_map[y][x] > 0:
                    print("%3d"%self.mine_map[y][x], end="")
                else:
                    print("%3s"%marks[self.game_board[y][x]], end="")
            print("")
        self.print_footer()

if __name__ == '__main__':
    b = Game()
    quitGame = False
    while not quitGame:
        b.print_game_board()
        print("o x y: セルを開く，f x y: フラグ設定/解除, q: 終了 -->", end="")
        command_str = input()

        try:
            cmd = command_str.split(" ")
            if cmd[0] == 'o':
                x, y = cmd[1:]
                if b.open_cell(int(x), int(y)) == False:
                    print("ゲームオーバー!")
                    quitGame = True
            elif cmd[0] == 'f':
                x, y = cmd[1:]
                b.flag_cell(int(x), int(y))
            elif cmd[0] == 'q':
                print("ゲームを終了します．")
                quitGame = True
                break
            else:
                print("コマンドはo, f, qのいずれかを指定してください．")
        except:
            print("もう一度，コマンドを入力してください．")
            
        if b.is_finished():
            b.print_game_board()
            print("ゲームクリア!")
            quitGame = True

File: minesweeper.c
#include<stdio.h>
#include<stdlib.h> // rand関数を使用するために必要
#include<time.h> // time関数を使用するために必要

#define MS_SIZE 8 // ゲームボードのサイズ(変更不可)
#define MINE -1 // 地雷のセル
#define FLAG 2 // フラグを設置したセル
#define OPEN 1 // 開いたセル

int main(void)
{
    int mode; // モードを放置 (モード：セルを開く、フラグを立てる/除去する)
    int number_of_mines; // 地雷数のデフォルト値
    int mine_map[MS_SIZE][MS_SIZE]; // 地雷セルと数字セルを記録
    int game_board[MS_SIZE][MS_SIZE]; // ゲームの進行を記録するためのゲームボード

    int count; // 立ったフラグの数を数える

    int again=1,choose; // ゲームもう一回やるについての設置
    
    srand((unsigned)time(NULL)); // time関数で現在時刻を取得し、乱数を初期化する
    
    // 課題1 課題2    
    int x,y; // 2次元配列で使うために
    int n=0; // 地雷かどうかを判断するために
    int i,j; // 2次元配列で使うため

    printf("\n%d× %dのボードの各セルに配置された地雷を除去するゲームです。\n", MS_SIZE, MS_SIZE);
    printf("ゲームを始めます。\n");

    while(again==1){
        printf("\n地雷数を設定してください。(1~%d)： ",(MS_SIZE*MS_SIZE)-1);
        scanf("%d", &number_of_mines);
        while(number_of_mines==0 || number_of_mines>=MS_SIZE*MS_SIZE){
	    printf("地雷数が正しくありません\n");
	    printf("地雷数を設定してください。(1~%d)： ",(MS_SIZE*MS_SIZE)-1);
	    scanf("%d", &number_of_mines);
        }

	printf("\n");

        // 初期化
        for(y=0; y<MS_SIZE; y++){

	    for(x=0; x<MS_SIZE; x++){
	        mine_map[y][x]=0;
	    }
        }

        // 地雷の位置をrand関数で決める
        for(i=0; i<number_of_mines; i++){
 	    y=rand()%MS_SIZE;
	    x=rand()%MS_SIZE;
            while(mine_map[y][x]==MINE){ // もう設定した場合
	        y=rand()%MS_SIZE;
	        x=rand()%MS_SIZE;
	    }
	    mine_map[y][x]=MINE;
        }
	
        // 各セルの8近傍の地雷をカウント
        for(y=0;y<MS_SIZE;y++){
            for(x=0;x<MS_SIZE;x++){
	        if(mine_map[y][x]==MINE){ // セルは地雷かどうかを判断する
	        }
	        else{
		    for(j=-1;j<2;j++){
		        for(i=-1;i<2;i++){
			    if(y+j<0 || x+i<0 || y+j>7 || x+i>7){
			    }
			    else{
			        if(mine_map[y+j][x+i]==MINE){ // 8近傍のセルは地雷かどうかを判断する
			            mine_map[y][x] +=1;
				}
			    }
			}
		    }
		}
	    }
	}
      
        // 課題3
        int a,b; // セルを選択するために
        int win=0; // ゲームを続けるかどうかを判断するために

        // 初期化する
        for(y=0;y<MS_SIZE;y++){
	    for(x=0;x<MS_SIZE;x++){
	       game_board[y][x]=0;
	    }
	}

        // 全体を表示する
        printf("[y]\n\n");

        for(y=0;y<MS_SIZE;y++){
  	    printf("%2d|",y);		
	    for(x=0;x<MS_SIZE;x++){
                printf("  x");
	    }
	    printf("\n");
        }

        printf("   ");
        for(x=0;x<MS_SIZE;x++){
	    printf("---");
        }	
        printf("\n");
        printf("   ");
        for(x=0;x<MS_SIZE;x++){
	    printf("%3d",x);
        }
        printf("  [x]\n\n");

        while(win==0){ // ゲームを続ける条件
            printf("モードを選択してください：セルを開く(1)、フラグを設置/除去する：(2)：");
	    scanf("%d",&mode);

	    while(mode!=1 && mode!=2){
	       printf("1か2を入力してください：");
	       scanf("%d",&mode);
	    }
		
	    // ユーザがセルを指定		
	    if(mode==1){
	        printf("セルを開きます。\n");
	        printf("[x]と[y]を入力してください。\n");
	        printf("[x]:");
	        scanf("%d",&a);
	        printf("[y]:");
	        scanf("%d",&b);
	        printf("\n");

	        while(a<0 || a>7 || b<0 || b>7){ // セル内かどうかを判断する
	            printf("0から7までの整数を入力してください。\n");
		    printf("[x]:");
		    scanf("%d",&a);
		    printf("[y]:");
		    scanf("%d",&b);
		    printf("\n");
	        }

	        if(mine_map[b][a]==MINE){ //地雷を選択した場合
		    printf("GAME OVER!\n");
		    win=1;

	        }

	        else{
		    game_board[b][a]=OPEN;
	            // 選択したセルの8近傍のセルを設定する
	            for(j=-1;j<2;j++){
		        for(i=-1;i<2;i++){
                            if(b+j<0 || a+i<0 || b+j>7 || a+i>7){
			    }
			    else{
                                if(game_board[b+j][a+i]==FLAG){
                                    game_board[b+j][a+i]=FLAG;
			        }
			        else{
			            game_board[b+j][a+i]=OPEN;
			        }
			    } 
			}
		    }
		}
	    }

	    // 課題4
	    else{ // フラグを設置/除去する
	        printf("フラグを設置/除去します。\n");
                printf("[x]と[y]を入力してください。\n");
	        printf("[x]:");
	        scanf("%d",&a);
	        printf("[y]:");
	        scanf("%d",&b);

	        while(a<0 || a>7 || b<0 || b>7){ // セル内か判断
		    printf("0から7までの整数を入力してください。\n");
	            printf("[x]:");
	            scanf("%d",&a);
	            printf("[y]:");
	            scanf("%d",&b);
	        }

	        if(game_board[b][a]==FLAG){ // グラフもう設置した場合
		    game_board[b][a]=OPEN;
	        }
	        else{
	            game_board[b][a]=FLAG;
	        }
	    }

            // 全体を表示する
	    if(win==0){
                printf("[y]\n\n");

                for(y=0;y<MS_SIZE;y++){
                    printf("%2d|",y);		
	            for(x=0;x<MS_SIZE;x++){
		        // 選択したセルの8近傍のセルだけを表示する
		        if(game_board[y][x]==OPEN){
                            if(mine_map[y][x]==0){ // 選択したセルの8近傍のセルが0の場合は何も表示しない
			        printf("   "); 
		            }
			    else if(mine_map[y][x]==MINE){ // 地雷のセル
			        printf("  x");
			    }
		            else{
			        printf("%3d",mine_map[y][x]);
		            }
		        }
		        else if(game_board[y][x]==FLAG){ // フラグを設置したセル
			    printf("  F");
		        }
		        else{
		            printf("  x");
		        }
	             }
	             printf("\n");
                }
                printf("   ");
                for(x=0;x<MS_SIZE;x++){
	            printf("---");
                }	
                printf("\n");
                printf("   ");
                for(x=0;x<MS_SIZE;x++){
	            printf("%3d",x);
                }
                printf("  [x]\n\n");
	    }
	    // 課題５
	    count=0;

	    // 開いたセルを数える
  	    for(y=0;y<MS_SIZE;y++){
	        for(x=0;x<MS_SIZE;x++){
		    if(game_board[y][x]==OPEN){
        	        count +=1;
	            }
                }
	    }
		   		
	    if(count==((MS_SIZE*MS_SIZE)-number_of_mines)){ // ゲームクリアの条件
	        printf("Congratulation!\n");

		win=1;
	    }
        }

        // 結果
        printf("\n*************結果*************\n\n");

        printf("[y]\n\n");

        for(y=0;y<MS_SIZE;y++){
	    printf("%2d|",y);		
            for(x=0;x<MS_SIZE;x++){
	        if(mine_map[y][x]==MINE){
		    printf("  M");
	        }
		else if(mine_map[y][x]==0){
		    printf("   ");
		}
	        else{
	            printf("%3d",mine_map[y][x]);
	        }
	    }
	    printf("\n");
        }

        printf("   ");
        for(x=0;x<MS_SIZE;x++){
	    printf("---");
        }	
        printf("\n");
        printf("   ");
        for(x=0;x<MS_SIZE;x++){
	    printf("%3d",x);
        }
        printf("  [x]\n\n");

        printf("ここで、Mは地雷です。\n\n");

        printf("ゲームを続行しますか？(Yes:1 No:2)");
        scanf("%d",&choose);

        again=choose;
    }


    return 0;    
}
